### DISCUSSION TITLE
let the GUI handle connections?

### DISCUSSION THREAD

**Discussion by umlaeute** [pure-data/pure-data#1767] [Ideas]
this is a sub-discussion for #1695 (*toolkit-agnostic Core->GUI Communication*)

in #1765 i experimented a bit with connection-handling on the GUI side.
the main idea was to mimick the `connect <srcobj> <outlet> <dstobj> <inlet>` message on the GUI side, and let the GUI do all the rest like:
- finding the correct outlet/inlet to connect
- use the correct connection type (message vs signal), based on information stored in the *outlet*
- automatically update connections whenever an object is moved (that is: re-position the connections)

### example

this patch:
```
#N canvas 0 0 450 300 12;
#X obj 57 113 print;
#X obj 57 88 unpack;
#X connect 1 1 0 0;
```

would translate to this code:

1. preparation (create two objects)
   ```
   ::pd::widget::create object 0x55d4482616a0 .x55d448258570.c 57 88
   ::pd::widget::config 0x55d4482616a0 -size 39 20 -text {unpack}
   ::pd::widget::create_inlets 0x55d4482616a0 {message}
   ::pd::widget::create_outlets 0x55d4482616a0 {message message}
   ::pd::widget::create object 0x55d44825d5e0 .x55d448258570.c 57 113
   ::pd::widget::config 0x55d44825d5e0 -size 39 20 -text {print}
   ::pd::widget::create_inlets 0x55d44825d5e0 {message}
   ```
   this creates objects with ID's `0x55d4482616a0` resp `0x55d44825d5e0` on the canvas `.x55d448258570.c`
the first object (`[unpack]` has a single message inlet and two message outlets.
the second object (`[print]`) has a single message inlet

2. connect the two objects:
   ``` 
   ::pd::widget::connect  0x55d4482616a0 1 0x55d44825d5e0 0
   ```

   this will use the 2nd outlet (`1`) of the first object (`0x55d4482616a0`) to find out its position (because this is where the connection starts) and its type (as this determines how the connection will look like).
it will also use the 1st inlet (`0`) of the second object (`0x55d44825d5e0`) to find out its position (as this is where the connection ends).
it will then create a connection cord.

3. do something with the objects:
   ```
   ::pd::widget::displace 0x55d4482616a0 100 0
   ```

   this will move the first object by 100 pixels to the right. the connection cord is updated accordingly

4. delete the connection
   ``` 
   ::pd::widget::disconnect 0x55d4482616a0 1 0x55d44825d5e0 0
   ```

5. removing an entire object would delete all the connections from this object as well:
   ```
   ::pd::widget::destroy 0x55d4482616a0
   ```

### cool, but.
this kind-of-works, but in the end i decided to ditch it.

the GUI obviously needs a way to determine where an iolet sits, whether it is a signal or a message iolet and to which object it belongs to.
unfortunately there is no standard way how GUI-objects create their iolets, so it's not really possible to extract that information from a random drawing on a canvas.

this means, that moving the connections to the GUI will simply break the connections of each and every existing GUI-external :smirk: 

### an intermediate step
so for now, the solution i came up with is to make "connections" simple widgets, that are actually independent of the objects they connect.
you basically tell the GUI to create a `connection` widget and then configure it to go from this position to that position.
(internally, this makes the `connection` widget only a thin wrapper around a `canvas line` (which could be modded to use splines and whatnot).

### where to go from here

if we actually want to head for the `::pd::widget::connect  0x55d4482616a0 1 0x55d44825d5e0 0`, we probably need a long-term plan on how to migrate their, without breaking existing GUI-externals.

some scenarios:
- provide a specification ("the new scheme") how iolets must be created so they can be used automatically; and hope that authors of GUI-externals will adopt
- provide procs to easily implement these specifications (allowing the authors some freedom on how the iolets actually look like); and hope that authors of GUI-externals will adopt
- just ignore existing GUI-externals and make them live without connections until their authors adopt the new scheme

afaict, we don't even have a way to check whether a given GUI-external has adopted the new scheme. we don't even know which items a GUI-external created, let alone whether it created any iolets (if we knew, we could spit out some big fat red warning, nagging people to adopt; and/or provide some fallback mechanism; but we don't).

if anybody has ideas how to solve this riddle: i'm all ears

**Comment by Ant1r** [comment]
> where to go from here

At first sight, I would have said `#3` (ignore and let existing GUI live without connection); 
of course it's a bit harsh, but it won't stop the patches from working... although it will surely make editing difficult.

Then I thought of a possible fix: 
- a `is_newgui` flag could be added somewhere (like in `t_text`, or in a 'private' area).
- internal objects and updated external ones would set `is_newgui` (preferably through a setter function so all of this could be removed in the future) at creation time.
- when calling the widgetbehavior functions, if `is_newgui` is not set, Pd would ask Tk to create/move/delete a fake invisible widget, that would be used for the connections. The dimensions of the fake widget would be obtained by calling the `w_parentgetrectfn` callback. The position of the iolets can be calculated knowing the dimensions and the number of iolets.

Do you think it could work?

**Reply by umlaeute** [reply]
that might indeed work.
new-style objects shouldn't need to set widgetbehaviours like `displacefn` (as displacing objects can now be handled entirely on the GUI), so this (or something similar) might be used to check whether its a legacy object or not.
